### PR TITLE
Allow usage of MiniMessage in server configuration files

### LINKED ISSUE #7520: Allow using MiniMessage in '...-message' configuration options
### Is your feature request related to a problem?

Since [Paper 1.18.2 will include Adventure 4.10.0](https://github.com/PaperMC/Paper/commit/9a73213f888b0dd9e4d3777068095747c6cd563d), MiniMessage will now also be included in Paper. Allowing the use of MiniMessage in the message config options would allow people to use hex colors with ease and use other features not available in the legacy text format, such as translatable components and clickable/hoverable text.

### Describe the solution you'd like.

A new config option (e.g. `messages.use-minimessage`), that when set to true serializes the messages in bukkit.yml, spigot.yml and paper.yml using MiniMessage instead of with the legacy ampersand serializer.

### Describe alternatives you've considered.

Creating gradients with legacy color codes (yuck)
image

### Other

_No response_

### PR DESCRIPTION
**Pull request by Astralchroma**
Allows the messages in `bukkit.yml`, `spigot.yml`, and `paper.yml` to be written using the MiniMessage format.

~~In order to maintain compatibility with existing configuration files the option `configuration-uses-minimessage` option has been added to `paper.yml`. This option defaults to `false` for any configuration files before config version 26.~~

New configuration files use MiniMessage by default, and the default messages have been converted to use it.

This also includes 2 minor API changes:
- `Bukkit#getPermissionMessage` which returned a `String` has been deprecated in favor of `Bukkit#permissionMessage` which returns a `Component`
- `Server#getPermissionMessage` which returned a `String` has been deprecated in favor of `Server#permissionMessage` which returns a `Component`

Closes #7520

### CODE REVIEW

File: patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
++++ b/src/main/java/org/bukkit/Bukkit.java
+@@ -2226,13 +2226,24 @@ public final class Bukkit {
+ 
+     /**
+      *
++     * @deprecated Use org.bukkit.Bukkit#permissionMessage()
+      * @return the default no permission message used on the server
+      */
+     @NotNull
++    @Deprecated
```
Comment (NoahvdAa) [inline patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: ~~Needs a `// Paper` comment, same for server~~

File: patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+             getDouble("settings.packet-limiter.limits.all.interval", 7.0),
+             getDouble("settings.packet-limiter.limits.all.max-packet-rate", 500.0),
+@@ -684,4 +735,9 @@ public class PaperConfig {
+     private static void useDimensionTypeForCustomSpawners() {
+         useDimensionTypeForCustomSpawners = getBoolean("settings.use-dimension-type-for-custom-spawners", false);
+     }
++
++    public static boolean configurationUsesMiniMessage;
++    private static void configurationUsesMiniMessage() {
++        configurationUsesMiniMessage = getBoolean("settings.configuration-uses-minimessage", version > 25);
```
Comment (MiniDigger) [inline patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Not sure I like this. How about instead, we check if the message contains ampersand, then revert to legacy, else default to MiniMessage

File: patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+     }
+ 
++    /**
++     *
++     * @return the default no permission message used on the server
++     */
++    @NotNull
++    public static net.kyori.adventure.text.Component permissionMessage() {
++        return server.permissionMessage();
++    }
```
Comment (NoahvdAa) [inline patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: ~~Needs `// Paper start` and `// Paper end`, same for server~~

File: patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+-        return msg != null ? io.papermc.paper.adventure.PaperAdventure.LEGACY_SECTION_UXRC.deserialize(msg) : null;
++        String msg = this.configuration.getString("settings.shutdown-message");
++
++        if (com.destroystokyo.paper.PaperConfig.configurationUsesMiniMessage)
++            // MiniMessage >> Component
++            return msg != null ? net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(msg) : null;
++
++        else
++            // Legacy >> Component
++            return msg != null ? io.papermc.paper.adventure.PaperAdventure.LEGACY_SECTION_UXRC.deserialize(msg) : null;
```
Comment (NoahvdAa) [inline patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Curly braces

File: patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch (hunk 2 above)
Comment (Astralchroma) [reply patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Just checking for usage of & could cause issues incase someone is using it other then formatting, having a config option is much more clear and leaves no room for it to get confused.

File: patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
++
++        if (string == null) {
++            authenticationServersDownKickMessage = null;
++
++        } else {
++            if (configurationUsesMiniMessage)
++                authenticationServersDownKickMessage = miniMessage().deserialize(string);
++
++            else
++                authenticationServersDownKickMessage = legacyAmpersand().deserialize(string);
```
Comment (NoahvdAa) [inline patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Same here and below, curly braces

File: patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch (hunk 2 above)
Comment (Machine-Maker) [reply patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch]: I don't think there should be a config option at all. And if the config version is < 26, all the messages should be force updated to minimessage by using the legacy and then minimessage serializers

**Review by NoahvdAa — COMMENTED** [review]
Congrats on making your first paper PR, welcome!

**Comment by electronicboy** [comment]
as I said on discord, really we should consilidate all of the messages into a singular patch file, rather than this patch which tweaks all of the old options, we'd also have to look into migrating bukkits messages over, but, am not sure 100% on that's based done without the config migration stuff

File: patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch (hunk 3 above)
Comment (Machine-Maker) [reply patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Its already inside a paper comment block

File: patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch (hunk 1 above)
Comment (Machine-Maker) [reply patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Already inside a paper comment block

File: patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch (hunk 3 above)
Comment (Astralchroma) [reply patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: They are already in a `//paper start` and `//paper end` block from another patch

File: patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
++    @Deprecated
+     String getPermissionMessage();
+ 
++    /**
++     *
++     * @return the default no permission message used on the server
++     */
++    @NotNull
++    net.kyori.adventure.text.Component permissionMessage();
++
```
Comment (Machine-Maker) [inline patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: I think this change (and the similar ones) should be made in the patch that added them.

File: patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+ 
++    /**
++     *
++     * @return the default no permission message used on the server
++     */
++    @NotNull
++    public static net.kyori.adventure.text.Component permissionMessage() {
++        return server.permissionMessage();
++    }
++
```
Comment (Machine-Maker) [inline patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Should be done in the patch that added this.

File: patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+diff --git a/src/main/java/org/bukkit/command/Command.java b/src/main/java/org/bukkit/command/Command.java
+index 95847fcb1ac2e430ee192f4f7ac94e981151c5db..9e2e4d8082b014229976f92c213b3084e09e11eb 100644
+--- a/src/main/java/org/bukkit/command/Command.java
++++ b/src/main/java/org/bukkit/command/Command.java
+@@ -185,7 +185,7 @@ public abstract class Command {
+         }
+ 
+             // Paper start - use components for permissionMessage
+-        net.kyori.adventure.text.Component permissionMessage = this.permissionMessage != null ? this.permissionMessage : io.papermc.paper.text.PaperComponents.legacySectionSerializer().deserialize(Bukkit.getPermissionMessage());
++        net.kyori.adventure.text.Component permissionMessage = this.permissionMessage != null ? this.permissionMessage : Bukkit.permissionMessage();
```
Comment (Machine-Maker) [inline patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Should be inside that patch that added it.

File: patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+ 
+     private static File CONFIG_FILE;
+@@ -71,8 +75,8 @@ public class PaperConfig {
+         commands.put("paper", new PaperCommand("paper"));
+         commands.put("mspt", new MSPTCommand("mspt"));
+ 
+-        version = getInt("config-version", 25);
+-        set("config-version", 25);
++        version = getInt("config-version", 26);
++        set("config-version", 26);
```
Comment (Machine-Maker) [inline patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch]: This change needs to be done in the very first config patch

File: patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
++        String kickVehicle = getString("messages.kick.flying-vehicle", "Flying is not enabled on this server");
++
++        if (configurationUsesMiniMessage) {
++            flyingKickPlayerMessage = miniMessage().deserialize(kickPlayer);
++            flyingKickVehicleMessage = miniMessage().deserialize(kickVehicle);
++
++        } else {
++            flyingKickPlayerMessage = legacyAmpersand().deserialize(kickPlayer);
++            flyingKickVehicleMessage = legacyAmpersand().deserialize(kickVehicle);
++        }
```
Comment (Machine-Maker) [inline patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Should be done in the patch that adds this.

File: patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
++            authenticationServersDownKickMessage = null;
++
++        } else {
++            if (configurationUsesMiniMessage)
++                authenticationServersDownKickMessage = miniMessage().deserialize(string);
++
++            else
++                authenticationServersDownKickMessage = legacyAmpersand().deserialize(string);
++        }
+     }
```
Comment (Machine-Maker) [inline patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Should be done in the patch that adds thsi.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
Ok, basically all of this should be done in the patch that makes these changes originally.

File: patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch (hunk 7 above)
Comment (Astralchroma) [reply patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: How do I find out what patch added something to begin with?

Comment (Astralchroma) [reply patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Mvm I think I found it.

File: patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch (hunk 6 above)
Comment (Astralchroma) [reply patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Moved.

File: patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch (hunk 7 above)
Comment (Astralchroma) [reply patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Moved.

File: patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch (hunk 9 above)
Comment (Peter-Crawley) [reply patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch:64]: Moved.

File: patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch (hunk 8 above)
Comment (Astralchroma) [reply patches/api/0373-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Moved.

File: patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch (hunk 10 above)
Comment (Peter-Crawley) [reply patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch:89]: Moved.

File: patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch (hunk 2 above)
Comment (Astralchroma) [reply patches/server/0880-Allow-usage-of-MiniMessage-in-server-configuration.patch]: Okay, I guess I will just force all of the messages up to MiniMessage then.

**Comment by Astralchroma** [comment]
> Ok, basically all of this should be done in the patch that makes these changes originally.

So I have been working on doing this, but when it comes to messages added by Spigot, that has to be done in a new patch, which is fine, but... patch 147 depends on one of these messages, so the patch has to go before 147.

How would I go about adding a patch before an existing patch?

**Comment by Machine-Maker** [comment]
> How would I go about adding a patch before an existing patch?

I wouldn't put it right before, I'd maybe put it in the adventure patch. @kashike might want to weigh in on that. but to add a patch, just interactive rebase to the commit you want to add a patch after, and then just make a new commit, and then continue rebasing.

**Comment by Astralchroma** [comment]
So after basically rewriting this PR, usage of MiniMessage is no longer optional, I will add code to automatically convert Legacy messages to MiniMessage in a bit.

I have also changed everything to be done as part of previous patches. *Which was very annoying and took forever.*

**Comment by Astralchroma** [comment]
*Not dead, just have alot of things to be doing, I will get back to this later.*

**Comment by Astralchroma** [comment]
Closed because I am not sure how to go about handling configuration migration, so I will leave someone else to deal with it.

**Pull request closed without merging**